### DISCUSSION TITLE
How to write generic, typed guards?

### DISCUSSION THREAD

**Discussion by corsali-albert** [statelyai/xstate#2310] [Q&A] [answered]
I'm trying to write a guard (using typescript) which can be used in multiple machines with different contexts. It's a very simple guard just to check for a specific keypress, but I'm new to xstate and not sure how to type this correctly. This probably isn't the best example because I could just define it inline and have the typing automatic, but I have other use cases with the same constraints so it's a good simple case.

What I've tried:

```ts
import { Guard } from "xstate";

import { MyContext } from "src/contexts";

const keyIsBackspace: Guard<MyContext, KeyboardEvent> = (
  _context: MyContext,
  event: KeyboardEvent
): boolean => event.key === "Backspace";

export { keyIsBackspace };
```

Unfortunately, that both isn't what I'm trying to do nor does it work: I don't want to use the specific `MyContext` (since it's to be used in multiple machines, and the context isn't relevant), and typescript complains that

```
Type '(_context: MyContext, event: KeyboardEvent) => boolean' is not assignable to type 'Guard<MyContext, KeyboardEvent>'.
  Type '(_context: MyContext, event: KeyboardEvent) => boolean' is missing the following properties from type 'GuardPredicate<MyContext, KeyboardEvent>': type, predicate
```

My best alternative at the moment is to define a function that does the leg work (something like `(event) => event.key === "Backspace"`) and import/use that in the inline guard definition, but that doesn't feel very clean to me and I'd generally like to define my guards separately just so the machine definition is less cluttered. Is there a better way to approach this?

**Comment by Andarist** [comment, marked as answer]
If you could share a codesandbox of the problem it would be really helpful.

I think that you don't have to use `Guard` type at all, just use a function:
```ts
const keyIsBackspace = (_context: {}, event: KeyboardEvent): boolean => event.key === "Backspace";
```

**Reply by corsali-albert** [reply]
Thanks for the reply!

That does work so I'll use it barring finding a better alternative, but my goal was to use a more purpose-specific version of the type (in case, e.g., the `Guard` type changes and I could return something other than a boolean in the future, or in case I wanted to look up the docs for it inline).

Here's an example of what I'm trying: https://codesandbox.io/s/shy-cherry-uh46b

**Reply by VanTanev** [reply]
The type you are looking for is [ConditionPredicate](https://github.com/davidkpiano/xstate/blob/507f2449837dcffd25b87ed229ada5787fb74bff/packages/core/src/types.ts#L119), not `Guard`.

**Reply by corsali-albert** [reply]
Amazing, thanks @VanTanev! Out of curiosity, what is the `Guard` type then?

**Reply by VanTanev** [reply]
It's an internal type representing a guard definition inside a machine definition.
